File: app/sim_client.py
"""Thin client for the Parking Simulator REST API (/api/v1).

Every protected call carries the bearer token. On a 401 the client logs in again
once and retries, so an expired token never takes the control loop down.

Note: the list-* endpoints carry a simulated operational cost. Call them once at
startup (or after a crash) to sync state, never on a polling loop.
"""
from urllib.parse import quote

import httpx

from app.config import settings


class SimError(Exception):
    pass


class SimClient:
    def __init__(self, base_url=None, user=None, password=None, timeout=None):
        self.base_url = (base_url or settings.sim_base_url).rstrip("/")
        self.user = user or settings.sim_user
        self.password = password or settings.sim_password
        self.token = None
        self.http = httpx.Client(timeout=timeout or settings.sim_timeout_s)

    # ---- auth -------------------------------------------------------------
    def login(self):
        r = self.http.post(f"{self.base_url}/auth/login",
                           json={"Email": self.user, "Password": self.password})
        if r.status_code != 200:
            raise SimError(f"login failed: {r.status_code} {r.text}")
        self.token = r.json()["token"]
        return self.token

    def _request(self, method, path, **kw):
        if self.token is None:
            self.login()
        url = f"{self.base_url}{path}"
        r = self.http.request(method, url, headers={"Authorization": f"Bearer {self.token}"}, **kw)
        if r.status_code == 401:
            self.login()
            r = self.http.request(method, url, headers={"Authorization": f"Bearer {self.token}"}, **kw)
        if r.status_code >= 400:
            raise SimError(f"{method} {path} -> {r.status_code} {r.text}")
        if not r.content:
            return None
        try:
            return r.json()
        except ValueError:
            return r.text

    def _get(self, path):
        return self._request("GET", path)

    def _post(self, path, **kw):
        return self._request("POST", path, **kw)

    # ---- discovery (costly: once per level load) ---------------------------
    def list_parking_spots(self): return self._get("/list-parking-spots")
    def list_barriers(self):      return self._get("/list-barriers")
    def list_lights(self):        return self._get("/list-lights")
    def list_exhaust_fans(self):  return self._get("/list-exhaust-fans")
    def list_alarms(self):        return self._get("/list-alarms")
    def list_zones(self):         return self._get("/list-zones")
    def test_webhook(self):       return self._get("/test")

    # ---- gates --------------------------------------------------------------
    def open_gate(self, name):   return self._post(f"/barrier-gates/{quote(name)}/open")
    def close_gate(self, name):  return self._post(f"/barrier-gates/{quote(name)}/close")
    def repair_gate(self, name): return self._post(f"/barrier-gates/{quote(name)}/repair")

    # ---- lights -------------------------------------------------------------
    def light_on(self, name):        return self._post(f"/lights/{quote(name)}/on")
    def light_off(self, name):       return self._post(f"/lights/{quote(name)}/off")
    def light_group_on(self, group):  return self._post(f"/lights/group/{quote(group)}/on")
    def light_group_off(self, group): return self._post(f"/lights/group/{quote(group)}/off")

    # ---- exhaust fans -------------------------------------------------------
    def fan_on(self, name):     return self._post(f"/exhaust-fans/{quote(name)}/on")
    def fan_off(self, name):    return self._post(f"/exhaust-fans/{quote(name)}/off")
    def repair_fan(self, name): return self._post(f"/exhaust-fans/{quote(name)}/repair")

    # ---- parking spots ------------------------------------------------------
    def repair_spot(self, name): return self._post(f"/parking-spots/{quote(name)}/repair")

    # ---- cars ---------------------------------------------------------------
    def car_goto(self, plate, destination):
        """destination: a parking spot name, 'exit' (pay at exit), or 'leavepark'."""
        return self._post(f"/car/{quote(plate)}/goto/{quote(destination)}")

    def car_charge(self, plate, parking_cost, charging_cost=0.0):
        """Issue the invoice. Must be sent exactly once, only while the car is at the exit spot."""
        return self._post(f"/car/{quote(plate)}/charge",
                          params={"parkingCost": parking_cost, "chargingCost": charging_cost})

File: server/src/simClient.ts
/**
 * Thin client for the Parking Simulator REST API (/api/v1).
 *
 * Every protected call carries the bearer token. On a 401 the client logs in again
 * once and retries, so an expired token never takes the control loop down.
 *
 * The list-* endpoints carry a simulated operational cost: call them once at startup
 * (or after a crash / level change) to sync state, never on a polling loop.
 */
import type { SimAlarm, SimBarrier, SimExhaustFan, SimLight, SimParkingSpot, SimZone } from "@gpa/shared";
import type { Settings } from "./config";

export class SimError extends Error {
  constructor(message: string, readonly outcomeUnknown = false) { super(message); }
}

/** What the controller and subsystems need from the simulator (a fake implements it in tests). */
export interface SimApi {
  listParkingSpots(): Promise<SimParkingSpot[]>;
  listBarriers(): Promise<SimBarrier[]>;
  listLights(): Promise<SimLight[]>;
  listExhaustFans(): Promise<SimExhaustFan[]>;
  listAlarms(): Promise<SimAlarm[]>;
  listZones(): Promise<SimZone[]>;
  openGate(name: string): Promise<void>;
  closeGate(name: string): Promise<void>;
  carGoto(plate: string, destination: string): Promise<void>;
  carCharge(plate: string, parkingCost: number, chargingCost: number): Promise<void>;
  repairGate(name: string): Promise<void>;
  repairSpot(name: string): Promise<void>;
  repairFan(name: string): Promise<void>;
  fanOn(name: string): Promise<void>;
  fanOff(name: string): Promise<void>;
  lightOn(name: string): Promise<void>;
  lightOff(name: string): Promise<void>;
  lightGroupOn(group: string): Promise<void>;
  lightGroupOff(group: string): Promise<void>;
}

export class SimClient implements SimApi {
  private token: string | null = null;
  private readonly baseUrl: string;

  constructor(private readonly cfg: Pick<Settings, "simBaseUrl" | "simUser" | "simPassword" | "simTimeoutS">) {
    this.baseUrl = cfg.simBaseUrl.replace(/\/$/, "");
  }

  // ---- auth -----------------------------------------------------------------
  async login(): Promise<string> {
    const r = await fetch(`${this.baseUrl}/auth/login`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ Email: this.cfg.simUser, Password: this.cfg.simPassword }),
      signal: AbortSignal.timeout(this.cfg.simTimeoutS * 1000),
    });
    if (r.status !== 200) throw new SimError(`login failed: ${r.status} ${await r.text()}`);
    this.token = ((await r.json()) as { token: string }).token;
    return this.token;
  }

  private async request<T>(method: "GET" | "POST", path: string): Promise<T> {
    if (!this.token) await this.login();
    const send = () =>
      fetch(`${this.baseUrl}${path}`, {
        method,
        headers: { authorization: `Bearer ${this.token}` },
        signal: AbortSignal.timeout(this.cfg.simTimeoutS * 1000),
      });
    let r: Response;
    try {
      r = await send();
    } catch (err) {
      throw new SimError(`${method} ${path} failed before a response: ${(err as Error).message}`, true);
    }
    if (r.status === 401) {
      await this.login();
      try {
        r = await send();
      } catch (err) {
        throw new SimError(`${method} ${path} failed before a response: ${(err as Error).message}`, true);
      }
    }
    const text = await r.text();
    if (r.status >= 400) throw new SimError(`${method} ${path} -> ${r.status} ${text}`);
    if (!text) return undefined as T;
    try {
      return JSON.parse(text) as T;
    } catch {
      return text as T;
    }
  }

  // ---- discovery (costly: once per level load) --------------------------------
  listParkingSpots = () => this.request<SimParkingSpot[]>("GET", "/list-parking-spots");
  listBarriers = () => this.request<SimBarrier[]>("GET", "/list-barriers");
  listLights = () => this.request<SimLight[]>("GET", "/list-lights");
  listExhaustFans = () => this.request<SimExhaustFan[]>("GET", "/list-exhaust-fans");
  listAlarms = () => this.request<SimAlarm[]>("GET", "/list-alarms");
  listZones = () => this.request<SimZone[]>("GET", "/list-zones");
  testWebhook = () => this.request<string>("GET", "/test");

  // ---- gates ------------------------------------------------------------------
  openGate = (n: string) => this.request<void>("POST", `/barrier-gates/${enc(n)}/open`);
  closeGate = (n: string) => this.request<void>("POST", `/barrier-gates/${enc(n)}/close`);
  repairGate = (n: string) => this.request<void>("POST", `/barrier-gates/${enc(n)}/repair`);

  // ---- lights -----------------------------------------------------------------
  lightOn = (n: string) => this.request<void>("POST", `/lights/${enc(n)}/on`);
  lightOff = (n: string) => this.request<void>("POST", `/lights/${enc(n)}/off`);
  lightGroupOn = (g: string) => this.request<void>("POST", `/lights/group/${enc(g)}/on`);
  lightGroupOff = (g: string) => this.request<void>("POST", `/lights/group/${enc(g)}/off`);

  // ---- exhaust fans -------------------------------------------------------------
  fanOn = (n: string) => this.request<void>("POST", `/exhaust-fans/${enc(n)}/on`);
  fanOff = (n: string) => this.request<void>("POST", `/exhaust-fans/${enc(n)}/off`);
  repairFan = (n: string) => this.request<void>("POST", `/exhaust-fans/${enc(n)}/repair`);

  // ---- parking spots ------------------------------------------------------------
  repairSpot = (n: string) => this.request<void>("POST", `/parking-spots/${enc(n)}/repair`);

  // ---- cars -----------------------------------------------------------------------
  /** destination: a parking spot name, "exit" (pay at exit), or "leavepark". */
  carGoto = (plate: string, destination: string) =>
    this.request<void>("POST", `/car/${enc(plate)}/goto/${enc(destination)}`);

  /** Issue the invoice. Exactly once, only while the car is at an exit spot. */
  carCharge = (plate: string, parkingCost: number, chargingCost = 0) =>
    this.request<void>("POST", `/car/${enc(plate)}/charge?parkingCost=${parkingCost}&chargingCost=${chargingCost}`);
}

const enc = encodeURIComponent;
